setup.py

```diff
@@ -17,11 +19,194 @@
-    setup_requires=['pbr']
-)
+
+ALL_STRING_TYPES = tuple(map(type, ('', b'', u'')))
+
+
+def str_if_nested_or_str(s):
+    """Turn input into a native string if possible."""
+    if isinstance(s, ALL_STRING_TYPES):
+        return str(s)
+    if isinstance(s, (list, tuple)):
+        return type(s)(map(str_if_nested_or_str, s))
+    if isinstance(s, (dict, )):
+        return stringify_dict_contents(s)
+    return s
+
+
+def stringify_dict_contents(dct):
+    """Turn dict keys and values into native strings."""
+    return {
+        str_if_nested_or_str(k): str_if_nested_or_str(v)
+        for k, v in dct.items()
+    }
+
+
+try:
+    from setuptools.config import read_configuration, ConfigOptionsHandler
+    import setuptools.config
+    import setuptools.dist
+
+    # Set default value for 'use_scm_version'
+    setattr(setuptools.dist.Distribution, 'use_scm_version', False)
+
+    # Attach bool parser to 'use_scm_version' option
+    class ShimConfigOptionsHandler(ConfigOptionsHandler):
+        """Extension class for ConfigOptionsHandler."""
+
+        @property
+        def parsers(self):
+            """Return an option mapping with default data type parsers."""
+            _orig_parsers = super(ShimConfigOptionsHandler, self).parsers
+            return dict(use_scm_version=self._parse_bool, **_orig_parsers)
+
+        def parse_section_packages__find(self, section_options):
+            find_kwargs = super(
+                ShimConfigOptionsHandler, self
+            ).parse_section_packages__find(section_options)
+            return stringify_dict_contents(find_kwargs)
+
+    setuptools.config.ConfigOptionsHandler = ShimConfigOptionsHandler
+except ImportError:
+    """This is a shim for setuptools<30.3."""
+    import io
+    import json
+
+    try:
+        from configparser import ConfigParser, NoSectionError
+    except ImportError:
+        from ConfigParser import ConfigParser, NoSectionError
+        ConfigParser.read_file = ConfigParser.readfp
+
+    def maybe_read_files(d):
+        """Read files if the string starts with `file:` marker."""
+        d = d.strip()
+        if not d.startswith('file:'):
+            return d
+        descs = []
+        for fname in map(str.strip, str(d[5:]).split(',')):
+            with io.open(fname, encoding='utf-8') as f:
+                descs.append(f.read())
+        return ''.join(descs)
+
+    def cfg_val_to_list(v):
+        """Turn config val to list and filter out empty lines."""
+        return list(filter(bool, map(str.strip, str(v).strip().splitlines())))
+
+    def cfg_val_to_dict(v):
+        """Turn config val to dict and filter out empty lines."""
+        return dict(
+            map(lambda l: list(map(str.strip, l.split('=', 1))),
+                filter(bool, map(str.strip, str(v).strip().splitlines())))
+        )
+
+    def cfg_val_to_primitive(v):
+        """Parse primitive config val to appropriate data type."""
+        return json.loads(v.strip().lower())
```

Comment:
Should probably be wrapped with a `try/except` block?

Reply:
No. It should fail on any errors. And again, it's not a full-featured implementation, it's just a shim. If it fails, it shouldn't hide this. Because in this case it'll eat all warning signs and we'll ship a broken sdist.